File: Chapter 04/JS/section_4_4_anti_patterns/after_rich_domain_thin_controller/data_access/dataAccessInterfaces.js
// Chapter 04/JavaScript/4.4 Anti-Patterns/After-RichDomain/data_access/dataAccessInterfaces.js

/**
 * ARCHITECTURE NOTE: In a traditional Layered Architecture, the 
 * Data Access Layer defines the contracts for accessing data. 
 * The Business Logic layer above will be forced to depend on 
 * this layer to use these interfaces.
 * * In JavaScript, we simulate interfaces by creating base classes 
 * that throw errors if their methods aren't overridden.
 */
class IOrderRepository {
    getById(orderId) { throw new Error("Not implemented"); }
    save(order) { throw new Error("Not implemented"); }
}

class ICustomerRepository {
    getById(customerId) { throw new Error("Not implemented"); }
}

class IEmailService {
    send(to, subject, body) { throw new Error("Not implemented"); }
}

module.exports = {
    IOrderRepository,
    ICustomerRepository,
    IEmailService
};

File: Chapter 04/C#/Section_4_4_Anti_Patterns/After_Rich_Domain_Thin_Controller/After/After.DataAccess/DataAccessInterfaces.cs
using After.DomainModels;

namespace After.DataAccess
{
    /// <summary>
    /// ARCHITECTURE NOTE: In a traditional N-Tier architecture, the 
    /// Data Access Layer defines the contracts for accessing data. 
    /// The Business Logic layer above will be forced to depend on 
    /// this layer to use these interfaces.
    /// </summary>
    public interface IOrderRepository
    {
        Order GetById(int orderId);
        void Save(Order order);
    }
    
    public interface ICustomerRepository
    {
        Customer GetById(int customerId);
    }
    
    public interface IEmailService
    {
        void Send(string to, string subject, string body);
    }
}